Refactor `post_process` in signature merging

clementblaudeau opened the pull request:
This (small) PR is a follow-up of #14100 and #14117. The `post_process` function for signature merging was taking two related arguments: 
* a `~destructive` boolean flag 
* a `replace` function to apply only if the flag was set to `true`

This PR combines the two into a single optional function `~replace` and clarifies some documentation comments.

clementblaudeau committed:
Refactor post_process of signature merging

The [post_process] function for signature merging was taking two linked
arguments: a [~destructive] flag and a [replace] function to apply only if the
flag was set to [true]. This commit combines the two into a single optional
function [replace] and clarifies some documentation comments.

Octachron commented on typing/typemod.ml:
```diff
@@ -554,27 +554,37 @@ module Merge = struct
+      phase of typing the signatures of recursive modules)
+
+      - [~invalid_alias] optional path, set when a module constraint is made
+      with a non-aliasable path to prevent the introduction of invalid
```
The "is made" part is not that clear to me. Maybe  "when the right hand side of a module constraint is a non-aliasable path" ?

clementblaudeau replied:
Rephrased in 706823f870

clementblaudeau committed:
Update comment and Changelog

Octachron commented on typing/typemod.ml:
```diff
@@ -770,8 +787,10 @@ module Merge = struct
+    (* There is no need to replace the constrained type, as all type fields are
+       made abstract anyway (so it should not appear anywhere) *)
+    let replace = None in
+    post_process ~approx:true ~replace loc lid env paths sg
```
`~replace:None` ?

clementblaudeau replied:
Done in b7a020fb50

Octachron commented on typing/typemod.ml:
```diff
@@ -716,6 +726,16 @@ module Merge = struct
+
     (* Post processing *)
     let replace =
       if destructive then
```
I propose to inverse the condition to have the easy case `not destructive` case first .

clementblaudeau replied:
Done in b7a020fb50 (also for module and module types constraints)

Octachron commented on typing/typemod.ml:
```diff
@@ -716,6 +726,16 @@ module Merge = struct
+       && destructive
+       && params_are_constrained params then
+      raise(Error(loc, env, With_cannot_remove_constrained_type));
+
```
The test ought to start with the conditions that requires no computations. Also introducing a name for just a condition feels a bit noisy to me. Overall, I still think this test is better localized in the corresponding branch of the `replace` function to avoid duplicating computations.

On the more nitpicky side, adding blank line inside a function makes it harder (for me) to identify the start and end of functions.

clementblaudeau replied:
I've moved the error computation back inside the replace function in b7a020fb50 (and removed empty lines)

clementblaudeau committed:
Address review comments

- Invert destructive and non-destructive cases
- Move the error inside the computation of the replace function

Octachron commented on typing/typemod.ml:
```diff
@@ -770,8 +781,9 @@ module Merge = struct
-    let replace = fun s _path -> s in
-    post_process ~approx:true ~destructive loc lid env paths sg replace
+    (* There is no need to replace the constrained type, as all type fields are
+       made abstract anyway (so it should not appear anywhere) *)
```
You should mention that we are also not checking the merged signature anyway because of `approx:true`.

clementblaudeau replied:
I've updated the comment in [a4a7be5](https://github.com/ocaml/ocaml/pull/14127/commits/a4a7be5ce4387cc91d9d30335c14d2be4d6921e8)

clementblaudeau committed:
Update comment

Octachron commented on typing/typemod.ml:
```diff
@@ -812,11 +825,13 @@ module Merge = struct
+    let replace =
+      if not destructive then None
+      else Some(fun s p -> Subst.Unsafe.add_module_path p path s)
+    in
```
Those lines make me wonder if we want a helper function, but then we would be back to a similar code than the pre-existing one. 
After pondering a bit, it is probably more locally visible with an explicit `if` when building the `replace` function.

Octachron reviewed (approved):
I agree that this PR makes the code more locally readable and is ready to merge.
